File: src/main/scala/story/Narrator.scala
package story

import Verbs._

class Narrator(val story: Story) {
  private val currentNode: NormalNode = story.startNode
  def start(): String = story.startNode.statement

  def input(response: String): String = {
    val words = response.split(' ').toList

    val verbOpt = findVerb(words)
    val nounOpt = findNoun(words)

    (verbOpt, nounOpt) match {
      case (Some(verb), Some(noun)) => handleMatch(verb, noun)
      case _                        => handleNoMatches(response)
    }
  }

  private def handleMatch(verb: Verb, noun: NounAndActions): String = {
    val nextNode = verb match {
      case Verbs.taste  => noun.taste
      case Verbs.touch  => noun.touch
      case Verbs.smell  => noun.smell
      case Verbs.look   => noun.look
      case Verbs.listen => noun.listen
    }
    nextNode match {
      case Some(NormalNode(statement, _)) => statement
      case Some(DeathNode(statement))     => s"$statement You died! x_x"
      case Some(WinNode(statement))       => s"$statement You won! °˖✧◝(⁰▿⁰)◜✧˖°"
      case _                              => s"You can't do that to a ${noun.item}."
    }
  }

  private def handleNoMatches(response: String) =
    response match {
      case "help" => "You need help?"
      case _      => "I don't understand"
    }

  private def findVerb(words: List[String]): Option[Verb] =
    Verbs.values.find(verb => words.contains(verb.toString()))

  private def findNoun(words: List[String]): Option[NounAndActions] =
    this.currentNode.stuffToDo.find(stuffToDo => words.contains(stuffToDo.item))

}

File: story-maker/src/chat/Narrator.ts
import { Story, Verb, VERBS, Node, StoryOption } from "../types";
import { head } from "ramda";

class Narrator {
  constructor(private story: Story, private currentNode: Node) {}

  public handleNonMatch = (response: string): string => {
    if (response.includes("help")) {
      return "TODO: helpful message.";
    }

    if (response.includes("look around")) {
      return this.lookAroundDescription();
    }

    return "I don't understand.";
  };

  public getNext = (response: string): Node | undefined => {
    const words = response.split(" ");

    const verbOpt = this.findVerb(words);
    const nounOpt = this.findNoun(words);

    if (verbOpt !== undefined && nounOpt !== undefined) {
      return this.handleMatch(verbOpt, nounOpt);
    } else {
      return undefined;
    }
  };

  private handleMatch = (verb: Verb, noun: StoryOption): Node | undefined => {
    switch (verb) {
      case "listen":
        return this.findNounById(noun.listenId);
      case "look":
        return this.findNounById(noun.lookId);
      case "smell":
        return this.findNounById(noun.smellId);
      case "taste":
        return this.findNounById(noun.tasteId);
      case "touch":
        return this.findNounById(noun.touchId);
    }
  };

  private findNounById = (id?: string) =>
    this.story.nodes.find(n => n.id === id);

  private stuffToDo = (): StoryOption[] =>
    this.story.options.filter(option =>
      this.currentNode.optionIds.includes(option.id)
    );

  private findVerb = (words: string[]): Verb | undefined =>
    VERBS.find(verb => words.includes(verb));

  private findNoun = (words: string[]): StoryOption | undefined =>
    this.stuffToDo().find(s => words.includes(s.item));

  private lookAroundDescription = (): string => {
    const items = this.stuffToDo()
      .map(stuff => `${aOrAn(stuff.item)} ${stuff.item}`)
      .join(" and ");
    return `You can see ${items}.`;
  };
}

const aOrAn = (word: string): string => {
  const vowels = "aeiou";
  const firstLetter = head(word);

  if (vowels.includes(firstLetter)) {
    return "an";
  }
  return "a";
};

export default Narrator;
